src/Altinn.App.Core/EFormidling/Implementation/DefaultEFormidlingService.cs

```diff
@@ -192,25 +192,47 @@ Instance instance
         return sbd;
     }
 
-    private async Task SendInstanceData(Instance instance, Dictionary<string, string> requestHeaders)
+    private async Task SendInstanceData(
+        Instance instance,
+        Dictionary<string, string> requestHeaders,
+        string eformidlingMetadataFilename
+    )
     {
         ApplicationMetadata applicationMetadata = await _appMetadata.GetApplicationMetadata();
 
         Guid instanceGuid = Guid.Parse(instance.Id.Split("/")[1]);
         int instanceOwnerPartyId = int.Parse(instance.InstanceOwner.PartyId, CultureInfo.InvariantCulture);
+
+        // Keep track of already used file names to ensure they are unique. eFormidling does not allow duplicate filenames.
+        HashSet<string> usedFileNames = [eformidlingMetadataFilename];
+
+        List<string> dataTypeIds = applicationMetadata.DataTypes.Select(x => x.Id).ToList();
+
         foreach (DataElement dataElement in instance.Data)
```

Comment:
Should we make this loop deterministic (sort or something)? AFAICT the `dataElement.Id` => `uniqueFilename` mapping can vary depending on the order of elements in `instance.Data`. `IServiceTask` may be retried in the future like, so idempotency and determinism will be important properties to have.

Otherwise LGTM 👍

Reply:
Simply order by id or something?
`foreach (DataElement dataElement in instance.Data.OrderBy(x => x.Id))`

Reply:
I think they get sorted by created time in storage.

I’m not sure when this order would matter, but it’s better to just sort.

Reply:
I _think_ that should work, I assume the sorting algo in .NET for strings is stable, so that the order is the same everytime

Reply:
Or is there a timestamp that could be used?

Reply:
Created datetime?

Reply:
Yeah that makes sense to me 👍